File: rti/internal/transport/grpc/support.go
// SupportService gRPC handler — translates rti.v1.SupportService RPCs
// into FOM-handle lookups. M25 Phase B (IEEE 1516.1-2010 §10.2).
//
// All methods are read-only against the per-federation FOM held by
// core.FOMRepository. The repository's Get returns a core.FOMHandle;
// the handler type-asserts that to core.FOMHandleNameLookup to access
// reverse + dimension lookups. Composition fault (production handle
// does not satisfy the lookup contract) returns codes.Internal.

package grpc

import (
	"context"
	"strings"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"

	"github.com/cbchoi/gorti/rti/internal/core"
	rtiv1 "github.com/cbchoi/gorti/rti/internal/genproto/rti/v1"
)

// Order-type / transport-type encodings per IEEE 1516.1-2010 §10.2
// (these are MIM-standard enum values; the SDK and the SupportService
// agree on the integer codes so the wire is symmetric).
const (
	OrderTypeReceive    uint32 = 1
	OrderTypeTimeStamp  uint32 = 2
	TransportReliable   uint32 = 1
	TransportBestEffort uint32 = 2
)

// supportService is the SupportServiceServer impl. Embeds the
// Unimplemented* shim per generated-code requirement.
type supportService struct {
	rtiv1.UnimplementedSupportServiceServer
	foms core.FOMRepository
}

func newSupportService(foms core.FOMRepository) *supportService {
	return &supportService{foms: foms}
}

// resolveLookup gets the FOM for a federation and asserts the reverse-
// lookup interface. Centralises the boilerplate that every handler
// needs.
func (s *supportService) resolveLookup(ctx context.Context, fed string) (core.FOMHandleNameLookup, error) {
	if strings.TrimSpace(fed) == "" {
		return nil, status.Error(codes.InvalidArgument, "federation_name is required")
	}
	h, err := s.foms.Get(ctx, core.FederationName(fed))
	if err != nil {
		return nil, status.Errorf(codes.NotFound, "federation %q: %v", fed, err)
	}
	nl, ok := h.(core.FOMHandleNameLookup)
	if !ok {
		return nil, status.Error(codes.Internal,
			"FOM handle does not implement reverse-lookup (composition fault)")
	}
	return nl, nil
}

func (s *supportService) GetObjectClassHandle(ctx context.Context, req *rtiv1.GetObjectClassHandleRequest) (*rtiv1.GetObjectClassHandleResponse, error) {
	if req == nil {
		return nil, nilRequest("GetObjectClassHandle")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	fh, ok := nl.(core.FOMHandle)
	if !ok {
		return nil, status.Error(codes.Internal, "FOM lookup missing forward-lookup contract")
	}
	h, ok := fh.LookupObjectClass(req.GetClassName())
	if !ok {
		return nil, status.Errorf(codes.NotFound, "object class %q not found", req.GetClassName())
	}
	return &rtiv1.GetObjectClassHandleResponse{ClassHandle: uint64(h)}, nil
}

func (s *supportService) GetObjectClassName(ctx context.Context, req *rtiv1.GetObjectClassNameRequest) (*rtiv1.GetObjectClassNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetObjectClassName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	name, ok := nl.ObjectClassName(core.ObjectClassHandle(req.GetClassHandle()))
	if !ok {
		return nil, status.Errorf(codes.NotFound, "object class handle %d not found", req.GetClassHandle())
	}
	return &rtiv1.GetObjectClassNameResponse{ClassName: name}, nil
}

func (s *supportService) GetAttributeHandle(ctx context.Context, req *rtiv1.GetAttributeHandleRequest) (*rtiv1.GetAttributeHandleResponse, error) {
	if req == nil {
		return nil, nilRequest("GetAttributeHandle")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	fh, ok := nl.(core.FOMHandle)
	if !ok {
		return nil, status.Error(codes.Internal, "FOM lookup missing forward-lookup contract")
	}
	h, ok := fh.LookupAttribute(core.ObjectClassHandle(req.GetClassHandle()), req.GetAttributeName())
	if !ok {
		return nil, status.Errorf(codes.NotFound,
			"attribute %q on class handle %d not found", req.GetAttributeName(), req.GetClassHandle())
	}
	return &rtiv1.GetAttributeHandleResponse{AttributeHandle: uint64(h)}, nil
}

func (s *supportService) GetAttributeName(ctx context.Context, req *rtiv1.GetAttributeNameRequest) (*rtiv1.GetAttributeNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetAttributeName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	name, ok := nl.AttributeName(
		core.ObjectClassHandle(req.GetClassHandle()),
		core.AttributeHandle(req.GetAttributeHandle()),
	)
	if !ok {
		return nil, status.Errorf(codes.NotFound,
			"attribute handle %d on class %d not found", req.GetAttributeHandle(), req.GetClassHandle())
	}
	return &rtiv1.GetAttributeNameResponse{AttributeName: name}, nil
}

func (s *supportService) GetInteractionClassHandle(ctx context.Context, req *rtiv1.GetInteractionClassHandleRequest) (*rtiv1.GetInteractionClassHandleResponse, error) {
	if req == nil {
		return nil, nilRequest("GetInteractionClassHandle")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	fh, ok := nl.(core.FOMHandle)
	if !ok {
		return nil, status.Error(codes.Internal, "FOM lookup missing forward-lookup contract")
	}
	h, ok := fh.LookupInteractionClass(req.GetClassName())
	if !ok {
		return nil, status.Errorf(codes.NotFound, "interaction class %q not found", req.GetClassName())
	}
	return &rtiv1.GetInteractionClassHandleResponse{ClassHandle: uint64(h)}, nil
}

func (s *supportService) GetInteractionClassName(ctx context.Context, req *rtiv1.GetInteractionClassNameRequest) (*rtiv1.GetInteractionClassNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetInteractionClassName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	name, ok := nl.InteractionClassName(core.InteractionClassHandle(req.GetClassHandle()))
	if !ok {
		return nil, status.Errorf(codes.NotFound, "interaction class handle %d not found", req.GetClassHandle())
	}
	return &rtiv1.GetInteractionClassNameResponse{ClassName: name}, nil
}

func (s *supportService) GetParameterHandle(ctx context.Context, req *rtiv1.GetParameterHandleRequest) (*rtiv1.GetParameterHandleResponse, error) {
	if req == nil {
		return nil, nilRequest("GetParameterHandle")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	fh, ok := nl.(core.FOMHandle)
	if !ok {
		return nil, status.Error(codes.Internal, "FOM lookup missing forward-lookup contract")
	}
	h, ok := fh.LookupParameter(core.InteractionClassHandle(req.GetClassHandle()), req.GetParameterName())
	if !ok {
		return nil, status.Errorf(codes.NotFound,
			"parameter %q on interaction class %d not found", req.GetParameterName(), req.GetClassHandle())
	}
	return &rtiv1.GetParameterHandleResponse{ParameterHandle: uint64(h)}, nil
}

func (s *supportService) GetParameterName(ctx context.Context, req *rtiv1.GetParameterNameRequest) (*rtiv1.GetParameterNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetParameterName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	name, ok := nl.ParameterName(
		core.InteractionClassHandle(req.GetClassHandle()),
		core.ParameterHandle(req.GetParameterHandle()),
	)
	if !ok {
		return nil, status.Errorf(codes.NotFound,
			"parameter handle %d on interaction class %d not found",
			req.GetParameterHandle(), req.GetClassHandle())
	}
	return &rtiv1.GetParameterNameResponse{ParameterName: name}, nil
}

func (s *supportService) GetDimensionHandle(ctx context.Context, req *rtiv1.GetDimensionHandleRequest) (*rtiv1.GetDimensionHandleResponse, error) {
	if req == nil {
		return nil, nilRequest("GetDimensionHandle")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	h, ok := nl.LookupDimension(req.GetDimensionName())
	if !ok {
		return nil, status.Errorf(codes.NotFound, "dimension %q not found", req.GetDimensionName())
	}
	return &rtiv1.GetDimensionHandleResponse{DimensionHandle: uint64(h)}, nil
}

func (s *supportService) GetDimensionName(ctx context.Context, req *rtiv1.GetDimensionNameRequest) (*rtiv1.GetDimensionNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetDimensionName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	name, ok := nl.DimensionName(core.DimensionHandle(req.GetDimensionHandle()))
	if !ok {
		return nil, status.Errorf(codes.NotFound, "dimension handle %d not found", req.GetDimensionHandle())
	}
	return &rtiv1.GetDimensionNameResponse{DimensionName: name}, nil
}

func (s *supportService) GetDimensionUpperBound(ctx context.Context, req *rtiv1.GetDimensionUpperBoundRequest) (*rtiv1.GetDimensionUpperBoundResponse, error) {
	if req == nil {
		return nil, nilRequest("GetDimensionUpperBound")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	nl, err := s.resolveLookup(ctx, req.GetFederationName())
	if err != nil {
		return nil, err
	}
	ub, ok := nl.DimensionUpperBound(core.DimensionHandle(req.GetDimensionHandle()))
	if !ok {
		return nil, status.Errorf(codes.NotFound, "dimension handle %d not found", req.GetDimensionHandle())
	}
	return &rtiv1.GetDimensionUpperBoundResponse{UpperBound: ub}, nil
}

// Order / transportation lookups — pure functions over the MIM enum,
// no FOM resolution required. Names are case-insensitive.
func (s *supportService) GetOrderType(_ context.Context, req *rtiv1.GetOrderTypeRequest) (*rtiv1.GetOrderTypeResponse, error) {
	if req == nil {
		return nil, nilRequest("GetOrderType")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	switch strings.ToLower(req.GetOrderName()) {
	case "receive":
		return &rtiv1.GetOrderTypeResponse{OrderType: OrderTypeReceive}, nil
	case "timestamp":
		return &rtiv1.GetOrderTypeResponse{OrderType: OrderTypeTimeStamp}, nil
	default:
		return nil, status.Errorf(codes.NotFound, "order name %q not recognised", req.GetOrderName())
	}
}

func (s *supportService) GetOrderName(_ context.Context, req *rtiv1.GetOrderNameRequest) (*rtiv1.GetOrderNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetOrderName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	switch req.GetOrderType() {
	case OrderTypeReceive:
		return &rtiv1.GetOrderNameResponse{OrderName: "Receive"}, nil
	case OrderTypeTimeStamp:
		return &rtiv1.GetOrderNameResponse{OrderName: "TimeStamp"}, nil
	default:
		return nil, status.Errorf(codes.NotFound, "order type %d not recognised", req.GetOrderType())
	}
}

func (s *supportService) GetTransportationType(_ context.Context, req *rtiv1.GetTransportationTypeRequest) (*rtiv1.GetTransportationTypeResponse, error) {
	if req == nil {
		return nil, nilRequest("GetTransportationType")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	switch strings.ToLower(req.GetTransportationName()) {
	case "hlareliable", "reliable":
		return &rtiv1.GetTransportationTypeResponse{TransportationType: TransportReliable}, nil
	case "hlabesteffort", "besteffort", "best-effort":
		return &rtiv1.GetTransportationTypeResponse{TransportationType: TransportBestEffort}, nil
	default:
		return nil, status.Errorf(codes.NotFound, "transportation name %q not recognised", req.GetTransportationName())
	}
}

func (s *supportService) GetTransportationName(_ context.Context, req *rtiv1.GetTransportationNameRequest) (*rtiv1.GetTransportationNameResponse, error) {
	if req == nil {
		return nil, nilRequest("GetTransportationName")
	}
	if err := validateWireVersion(req.GetWireVersion()); err != nil {
		return nil, err
	}
	switch req.GetTransportationType() {
	case TransportReliable:
		return &rtiv1.GetTransportationNameResponse{TransportationName: "HLAreliable"}, nil
	case TransportBestEffort:
		return &rtiv1.GetTransportationNameResponse{TransportationName: "HLAbestEffort"}, nil
	default:
		return nil, status.Errorf(codes.NotFound, "transportation type %d not recognised", req.GetTransportationType())
	}
}

File: pysdk/rti1516e/support.py
"""Layer-1 SDK client for the SupportService (M25 Phase B).

Thin async wrapper around the generated ``SupportServiceStub`` exposing
IEEE 1516.1-2010 §10.2 handle / name / dimension / order / transport
lookup services.

Composition mirrors the cut-3 service-group clients (sync/ddm/...):

    async with RtiConnection.connect("grpc://...") as rti:
        async with rti.join_federation(spec, federate_name="alice") as fed:
            vehicle_h = await fed.support.get_object_class_handle("Vehicle")
            pos_h = await fed.support.get_attribute_handle(vehicle_h, "Position")

The Layer-2 ``Rti1516eAmbassador`` exposes equivalent sync wrappers
(``getObjectClassHandle`` / ``getObjectClassName`` / ...) per reference_rti
naming for federates ported from commercial RTIs.

All lookups are read-only against the federation's FOM and do not
mutate federate state.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from rti1516e._grpc_errors import translate_rpc_error

if TYPE_CHECKING:  # pragma: no cover - type-check imports only
    import grpc


# Mirror the Go-side constants in rti/internal/transport/grpc/support.go.
ORDER_TYPE_RECEIVE = 1
ORDER_TYPE_TIMESTAMP = 2
TRANSPORT_RELIABLE = 1
TRANSPORT_BEST_EFFORT = 2


class SupportClient:
    """Federate-bound client for the SupportService gRPC surface.

    Construct via :meth:`Federate.support`; every method auto-populates
    ``wire_version=WIRE_VERSION_V1`` and the federation_name the
    federate joined under. The class_handle / interaction_class_handle
    parameters are passed by handle (not name) — call
    :meth:`get_object_class_handle` first to look one up.
    """

    def __init__(
        self,
        channel: grpc.aio.Channel,
        *,
        federation_name: str,
    ) -> None:
        from rti.v1 import support_pb2_grpc

        self._stub = support_pb2_grpc.SupportServiceStub(channel)
        self._federation_name = federation_name

    async def get_object_class_handle(self, class_name: str) -> int:
        """§10.2 — resolve an object class name to its handle."""
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetObjectClassHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_name=class_name,
        )
        try:
            resp = await self._stub.GetObjectClassHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.class_handle)

    async def get_object_class_name(self, class_handle: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetObjectClassNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
        )
        try:
            resp = await self._stub.GetObjectClassName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.class_name)

    async def get_attribute_handle(self, class_handle: int, attribute_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetAttributeHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
            attribute_name=attribute_name,
        )
        try:
            resp = await self._stub.GetAttributeHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.attribute_handle)

    async def get_attribute_name(self, class_handle: int, attribute_handle: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetAttributeNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
            attribute_handle=int(attribute_handle),
        )
        try:
            resp = await self._stub.GetAttributeName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.attribute_name)

    async def get_interaction_class_handle(self, class_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetInteractionClassHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_name=class_name,
        )
        try:
            resp = await self._stub.GetInteractionClassHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.class_handle)

    async def get_interaction_class_name(self, class_handle: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetInteractionClassNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
        )
        try:
            resp = await self._stub.GetInteractionClassName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.class_name)

    async def get_parameter_handle(self, class_handle: int, parameter_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetParameterHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
            parameter_name=parameter_name,
        )
        try:
            resp = await self._stub.GetParameterHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.parameter_handle)

    async def get_parameter_name(self, class_handle: int, parameter_handle: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetParameterNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            class_handle=int(class_handle),
            parameter_handle=int(parameter_handle),
        )
        try:
            resp = await self._stub.GetParameterName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.parameter_name)

    async def get_dimension_handle(self, dimension_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetDimensionHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            dimension_name=dimension_name,
        )
        try:
            resp = await self._stub.GetDimensionHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.dimension_handle)

    async def get_dimension_name(self, dimension_handle: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetDimensionNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            dimension_handle=int(dimension_handle),
        )
        try:
            resp = await self._stub.GetDimensionName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.dimension_name)

    async def get_dimension_upper_bound(self, dimension_handle: int) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetDimensionUpperBoundRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            dimension_handle=int(dimension_handle),
        )
        try:
            resp = await self._stub.GetDimensionUpperBound(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.upper_bound)

    async def get_order_type(self, order_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetOrderTypeRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            order_name=order_name,
        )
        try:
            resp = await self._stub.GetOrderType(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.order_type)

    async def get_order_name(self, order_type: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetOrderNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            order_type=int(order_type),
        )
        try:
            resp = await self._stub.GetOrderName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.order_name)

    async def get_transportation_type(self, transportation_name: str) -> int:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetTransportationTypeRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            transportation_name=transportation_name,
        )
        try:
            resp = await self._stub.GetTransportationType(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.transportation_type)

    async def get_transportation_name(self, transportation_type: int) -> str:
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetTransportationNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            transportation_type=int(transportation_type),
        )
        try:
            resp = await self._stub.GetTransportationName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.transportation_name)

    async def get_object_instance_handle(self, object_name: str) -> int:
        """§6.30 — resolve a runtime object instance name to its handle.

        M27 Phase C. Raises if no instance with that name is registered.
        """
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetObjectInstanceHandleRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            object_name=object_name,
        )
        try:
            resp = await self._stub.GetObjectInstanceHandle(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return int(resp.object_handle)

    async def get_object_instance_name(self, object_handle: int) -> str:
        """§6.31 — resolve a runtime object instance handle to its name."""
        from rti.v1 import common_pb2, support_pb2

        req = support_pb2.GetObjectInstanceNameRequest(
            wire_version=common_pb2.WireVersion.WIRE_VERSION_V1,
            federation_name=self._federation_name,
            object_handle=int(object_handle),
        )
        try:
            resp = await self._stub.GetObjectInstanceName(req)
        except Exception as exc:  # noqa: BLE001
            translate_rpc_error(exc)
            raise
        return str(resp.object_name)
